File: packages/engine-python/src/nomark_engine/schema.py
"""Schema types and validation for NOMARK ledger entries.

Ports packages/engine/src/schema.ts — Pydantic v2 models instead of Zod.
"""

from __future__ import annotations

import re
from typing import Literal, Union

from pydantic import BaseModel, Field, field_validator

# --- String literal unions (match TS types) ---

Context = Literal["chat", "cowork", "code"]
Outcome = Literal["accepted", "edited", "corrected", "rejected", "abandoned", "unknown"]
RequestType = Literal["question", "task", "brainstorm", "decision", "critique", "creative", "continuation", "reaction"]
PatternType = Literal["rewrite_request", "scope_change", "quality_complaint", "format_request", "style_override", "abort"]
RubricStage = Literal["ephemeral", "pending", "proven", "trusted"]
SignalType = Literal["meta", "pref", "map", "asn", "rub"]

# Scope is a free-form string: '*' | 'context:{ctx}' | 'topic:{topic}' | compound
Scope = str

_ISO_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_ISO_DATETIME_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}")


def _validate_iso_date(v: str) -> str:
    if not _ISO_DATE_RE.match(v):
        raise ValueError(f"Expected ISO date (YYYY-MM-DD), got {v!r}")
    return v


# --- Context counts ---

class ContextCounts(BaseModel):
    chat: int | None = None
    code: int | None = None
    cowork: int | None = None


# --- Signal data models ---

class SigPref(BaseModel):
    dim: str
    target: str
    w: float = Field(ge=0, le=1)
    n: int = Field(ge=0)
    src: ContextCounts
    ctd: int = Field(ge=0)
    scope: str
    decay: float = Field(ge=0, le=1)
    last: str
    staged: bool | None = None
    note: str | None = None

    @field_validator("last")
    @classmethod
    def _check_last(cls, v: str) -> str:
        return _validate_iso_date(v)


class SigMap(BaseModel):
    trigger: str
    pattern_type: PatternType
    intent: list[str]
    neg: list[str] | None = None
    conf: float = Field(ge=0, le=1)
    n: int = Field(ge=0)
    scope: str | None = None
    last: str

    @field_validator("last")
    @classmethod
    def _check_last(cls, v: str) -> str:
        return _validate_iso_date(v)


class SigAsn(BaseModel):
    field: str
    default: str
    accuracy: float = Field(ge=0, le=1)
    total: int = Field(ge=0)
    correct: int = Field(ge=0)
    last: str

    @field_validator("last")
    @classmethod
    def _check_last(cls, v: str) -> str:
        return _validate_iso_date(v)


class OutcomeCounts(BaseModel):
    accepted: int | None = None
    edited: int | None = None
    corrected: int | None = None
    rejected: int | None = None
    abandoned: int | None = None


class SigMeta(BaseModel):
    profile: dict[str, object]
    signals: int = Field(ge=0)
    by_ctx: ContextCounts
    by_out: OutcomeCounts
    avg_conf: float = Field(ge=0, le=1)
    avg_q: float = Field(ge=0)
    updated: str

    @field_validator("updated")
    @classmethod
    def _check_updated(cls, v: str) -> str:
        return _validate_iso_date(v)


class SigRub(BaseModel):
    id: str
    fmt: str
    stage: RubricStage
    uses: int = Field(ge=0)
    accepts: int = Field(ge=0)
    avg_ed: float = Field(ge=0)
    dims: dict[str, float]
    min: float
    ref: str | None = None
    last: str

    @field_validator("last")
    @classmethod
    def _check_last(cls, v: str) -> str:
        return _validate_iso_date(v)


# --- Ledger entry (discriminated union) ---

class LedgerEntryMeta(BaseModel):
    type: Literal["meta"] = "meta"
    data: SigMeta


class LedgerEntryPref(BaseModel):
    type: Literal["pref"] = "pref"
    data: SigPref


class LedgerEntryMap(BaseModel):
    type: Literal["map"] = "map"
    data: SigMap


class LedgerEntryAsn(BaseModel):
    type: Literal["asn"] = "asn"
    data: SigAsn


class LedgerEntryRub(BaseModel):
    type: Literal["rub"] = "rub"
    data: SigRub


LedgerEntry = Union[LedgerEntryMeta, LedgerEntryPref, LedgerEntryMap, LedgerEntryAsn, LedgerEntryRub]

_SIGNAL_TYPE_MAP: dict[str, type[BaseModel]] = {
    "meta": LedgerEntryMeta,
    "pref": LedgerEntryPref,
    "map": LedgerEntryMap,
    "asn": LedgerEntryAsn,
    "rub": LedgerEntryRub,
}

_SIGNAL_DATA_MAP: dict[str, type[BaseModel]] = {
    "meta": SigMeta,
    "pref": SigPref,
    "map": SigMap,
    "asn": SigAsn,
    "rub": SigRub,
}


def parse_ledger_entry(signal_type: str, data: dict) -> LedgerEntry | None:
    """Parse a signal type + data dict into a typed LedgerEntry. Returns None on validation failure."""
    entry_cls = _SIGNAL_TYPE_MAP.get(signal_type)
    if entry_cls is None:
        return None
    try:
        return entry_cls.model_validate({"type": signal_type, "data": data})  # type: ignore[return-value]
    except Exception:
        return None

File: packages/engine/src/schema.ts
import { z } from 'zod'

// --- Enums as string literal unions ---

export type Context = 'chat' | 'cowork' | 'code'
export type Outcome = 'accepted' | 'edited' | 'corrected' | 'rejected' | 'abandoned' | 'unknown'
export type RequestType = 'question' | 'task' | 'brainstorm' | 'decision' | 'critique' | 'creative' | 'continuation' | 'reaction'
export type PatternType = 'rewrite_request' | 'scope_change' | 'quality_complaint' | 'format_request' | 'style_override' | 'abort' | 'custom'
export type RubricStage = 'ephemeral' | 'pending' | 'proven' | 'trusted'
export type SignalType = 'meta' | 'pref' | 'map' | 'asn' | 'rub'

export type Scope = string // '*' | 'context:{ctx}' | 'topic:{topic}' | compound 'context:{ctx}+topic:{topic}'

// --- Zod schemas ---

const IsoDatePattern = /^\d{4}-\d{2}-\d{2}$/
const IsoDatetimePattern = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/

const IsoDate = z.string().regex(IsoDatePattern)
const IsoDatetime = z.string().regex(IsoDatetimePattern)

const ContextSchema = z.enum(['chat', 'cowork', 'code'])
const OutcomeSchema = z.enum(['accepted', 'edited', 'corrected', 'rejected', 'abandoned', 'unknown'])
const RequestTypeSchema = z.enum(['question', 'task', 'brainstorm', 'decision', 'critique', 'creative', 'continuation', 'reaction'])
const PatternTypeSchema = z.enum(['rewrite_request', 'scope_change', 'quality_complaint', 'format_request', 'style_override', 'abort', 'custom'])
const RubricStageSchema = z.enum(['ephemeral', 'pending', 'proven', 'trusted'])

const ContextCountsSchema = z.object({
  chat: z.number().int().optional(),
  code: z.number().int().optional(),
  cowork: z.number().int().optional(),
})

export const SigPrefSchema = z.object({
  dim: z.string(),
  target: z.string(),
  w: z.number().min(0).max(1),
  n: z.number().int().min(0),
  src: ContextCountsSchema,
  ctd: z.number().int().min(0),
  scope: z.string(),
  decay: z.number().min(0).max(1),
  last: IsoDate,
  staged: z.boolean().optional(),
  note: z.string().optional(),
  source_quote: z.string().optional(),
  source_scope: z.string().optional(),
})

export const SigMapSchema = z.object({
  trigger: z.string(),
  pattern_type: PatternTypeSchema,
  intent: z.array(z.string()),
  neg: z.array(z.string()).optional(),
  conf: z.number().min(0).max(1),
  n: z.number().int().min(0),
  scope: z.string().optional(),
  last: IsoDate,
})

export const SigAsnSchema = z.object({
  field: z.string(),
  default: z.string(),
  accuracy: z.number().min(0).max(1),
  total: z.number().int().min(0),
  correct: z.number().int().min(0),
  last: IsoDate,
})

export const SigMetaSchema = z.object({
  profile: z.record(z.string(), z.unknown()),
  signals: z.number().int().min(0),
  by_ctx: ContextCountsSchema,
  by_out: z.object({
    accepted: z.number().int().optional(),
    edited: z.number().int().optional(),
    corrected: z.number().int().optional(),
    rejected: z.number().int().optional(),
    abandoned: z.number().int().optional(),
  }),
  avg_conf: z.number().min(0).max(1),
  avg_q: z.number().min(0),
  updated: IsoDate,
})

export const SigRubSchema = z.object({
  id: z.string(),
  fmt: z.string(),
  stage: RubricStageSchema,
  uses: z.number().int().min(0),
  accepts: z.number().int().min(0),
  avg_ed: z.number().min(0),
  dims: z.record(z.string(), z.number().min(0).max(1)),
  min: z.number(),
  ref: z.string().optional(),
  last: IsoDate,
})

export const ArchiveEventSchema = z.object({
  id: z.string().regex(/^sig-/),
  timestamp: IsoDatetime,
  context: ContextSchema,
  session_id: z.string(),
  input_tier: z.number().int().min(0).max(2),
  request_type: RequestTypeSchema.optional(),
  input_summary: z.string().optional(),
  intent_confidence: z.number().min(0).max(1).optional(),
  assumptions: z.array(z.object({
    field: z.string(),
    value: z.string(),
    source: z.string(),
    correct: z.boolean().optional(),
  })).optional(),
  questions_asked: z.number().int().min(0).optional(),
  output_type: z.string().optional(),
  rubric_id: z.string().optional(),
  self_eval_score: z.number().optional(),
  outcome: OutcomeSchema,
  outcome_confidence: z.number().min(0).max(1).optional(),
  edit_distance: z.number().min(0).optional(),
  correction: z.string().nullable().optional(),
  resolver_decisions: z.array(z.object({
    dim: z.string(),
    winner: z.string(),
    score: z.number(),
    runner_up: z.string().optional(),
  })).optional(),
  preference_updates: z.array(z.object({
    dim: z.string(),
    delta: z.number(),
    reason: z.string().optional(),
  })).optional(),
})

export const LedgerEntrySchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('meta'), data: SigMetaSchema }),
  z.object({ type: z.literal('pref'), data: SigPrefSchema }),
  z.object({ type: z.literal('map'), data: SigMapSchema }),
  z.object({ type: z.literal('asn'), data: SigAsnSchema }),
  z.object({ type: z.literal('rub'), data: SigRubSchema }),
])

// --- Inferred types ---

export type SigPref = z.infer<typeof SigPrefSchema>
export type SigMap = z.infer<typeof SigMapSchema>
export type SigAsn = z.infer<typeof SigAsnSchema>
export type SigMeta = z.infer<typeof SigMetaSchema>
export type SigRub = z.infer<typeof SigRubSchema>
export type ArchiveEvent = z.infer<typeof ArchiveEventSchema>
export type LedgerEntry = z.infer<typeof LedgerEntrySchema>
